File: src/domain/geo.ts
import type { GeoPoint } from "./types";

const EARTH_RADIUS_KM = 6371;

export function distanceKmBetween(from: GeoPoint, to: GeoPoint): number {
  const latDelta = toRadians(to.lat - from.lat);
  const lngDelta = toRadians(to.lng - from.lng);
  const fromLat = toRadians(from.lat);
  const toLat = toRadians(to.lat);

  const haversine =
    Math.sin(latDelta / 2) ** 2 +
    Math.cos(fromLat) * Math.cos(toLat) * Math.sin(lngDelta / 2) ** 2;

  return Math.round(2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(haversine)) * 10) / 10;
}

function toRadians(value: number): number {
  return (value * Math.PI) / 180;
}

File: CarRentalOptimizer/Domain/Geo.swift
import Foundation

private let earthRadiusKm: Double = 6371

func distanceKmBetween(_ from: GeoPoint, _ to: GeoPoint) -> Double {
    let latDelta = toRadians(to.lat - from.lat)
    let lngDelta = toRadians(to.lng - from.lng)
    let fromLat = toRadians(from.lat)
    let toLat = toRadians(to.lat)

    let haversine = sin(latDelta / 2) * sin(latDelta / 2)
        + cos(fromLat) * cos(toLat) * sin(lngDelta / 2) * sin(lngDelta / 2)

    return (2 * earthRadiusKm * asin(sqrt(haversine)) * 10).rounded() / 10
}

private func toRadians(_ value: Double) -> Double {
    value * .pi / 180
}
